agent/immortalstreams/handler.go

```diff
@@ -0,0 +1,283 @@
+		switch {
+		case errors.Is(err, ErrTooManyStreams):
+			httpapi.Write(ctx, w, http.StatusServiceUnavailable, codersdk.Response{
+				Message: "Too many Immortal Streams.",
+			})
+			return
+		case errors.Is(err, ErrConnRefused):
+			httpapi.Write(ctx, w, http.StatusNotFound, codersdk.Response{
+				Message: "The connection was refused.",
+			})
+			return
+		default:
+			httpapi.InternalServerError(w, err)
+			return
+		}
+	}
+
+	httpapi.Write(ctx, w, http.StatusCreated, stream)
+}
+
+// listStreams lists all immortal streams
+func (h *Handler) listStreams(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	streams := h.manager.ListStreams()
+	httpapi.Write(ctx, w, http.StatusOK, streams)
+}
+
+// handleStreamRequest handles GET requests for a specific stream and returns stream info or handles WebSocket upgrades
+func (h *Handler) handleStreamRequest(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+
+	// Handle websocket upgrade for clients that use it
+	if httpapi.IsWebsocketUpgrade(r) {
+		h.handleUpgrade(w, r)
+		return
+	}
+
+	// No upgrade: return stream info
+	streamID := getStreamID(ctx)
+	stream, ok := h.manager.GetStream(streamID)
+	if !ok {
+		httpapi.Write(ctx, w, http.StatusNotFound, codersdk.Response{
+			Message: "Stream not found",
+		})
+		return
+	}
+
+	httpapi.Write(ctx, w, http.StatusOK, stream.ToAPI())
+}
+
+// deleteStream deletes a stream
+func (h *Handler) deleteStream(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	streamID := getStreamID(ctx)
+
+	err := h.manager.DeleteStream(streamID)
+	if err != nil {
+		switch {
+		case errors.Is(err, ErrStreamNotFound):
+			httpapi.Write(ctx, w, http.StatusNotFound, codersdk.Response{
+				Message: "Stream not found",
+			})
+			return
+		default:
+			httpapi.InternalServerError(w, err)
+			return
+		}
+	}
+
+	w.WriteHeader(http.StatusNoContent)
+}
+
+// handleUpgrade handles WebSocket upgrade for immortal stream connections
+func (h *Handler) handleUpgrade(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	streamID := getStreamID(ctx)
+
+	// Get sequence numbers from headers
+	readSeqNum, err := parseSequenceNumber(r.Header.Get(codersdk.HeaderImmortalStreamSequenceNum))
+	if err != nil {
+		httpapi.Write(ctx, w, http.StatusBadRequest, codersdk.Response{
+			Message: fmt.Sprintf("Invalid sequence number: %v", err),
+		})
+		return
+	}
+
+	// Check if stream exists
+	stream, ok := h.manager.GetStream(streamID)
+	if !ok {
+		httpapi.Write(ctx, w, http.StatusNotFound, codersdk.Response{
+			Message: "Stream not found",
+		})
+		return
+	}
+
+	// Send the current remote reader sequence number as a response header
+	// so clients using a backed pipe can resume their writer accurately.
+	if stream != nil && stream.GetPipe() != nil {
+		readerSeqNum := stream.GetPipe().ReaderSequenceNum()
+		w.Header().Set(codersdk.HeaderImmortalStreamSequenceNum, strconv.FormatUint(readerSeqNum, 10))
+	}
+
+	// Upgrade to WebSocket
```

Comment:
This is supposed to be an upgrade to `upgrade: coder-immortal-stream`

https://www.notion.so/coderhq/Immortal-Streams-229d579be59280b8a59defa327df384e?source=copy_link#229d579be592801489d3e8b28dcbe2b4

Not `upgrade: websocket` with `coder-immortal-stream` as the subprotocol.

Websockets just add overhead here: we want a raw TCP connection.  The reason we use websockets elsewhere is that they play more nicely with reverse proxies. But here we are going inside an end-to-end encrypted tunnel, so we know there are no HTTP proxies on the path.